### DISCUSSION TITLE
Native Image Committer and Community Meeting 2023-06-29

### DISCUSSION THREAD

**Discussion by christianwimmer** [oracle/graal#6906] [Show and tell]
List of all past and upcoming meetings: #3933

## New and Noteworthy

* Simulation of class initializer is now merged:
  #6710
  Now we are working to enable the new class initalization strategy by default.
  
* Starting JDK 22 work early this time:
  #6803
  #6879

* Reduce the module dependencies of the image generator (for cleanliness reasons):
  #6807
  #6901

* Small steps to make image builds more deterministic:
  #6900
  
* Performance
  #6765
  
* Compatibility: 
  #6835
  
* Monitoring / tools:
  #6774
  The old "enterprise" Dwarf information generation is now deleted from our code base: #6665
  #6705

**Comment by teshull** [comment]
## Runtime Compilation Support

---

## ParseOnce

* (CWimmer) Added ParseOnce feature in [May 2021](https://github.com/oracle/graal/commit/e92dab09)
    - Same graph used to create analysis flows also is used for compilation
    - Enables optimizations to be performed before analysis

* Allowed Spring to run on Native-Image

* Still not enabled for applications which perform runtime compilation (i.e., Truffle languages)

---

## ParseOnceJIT

* Goal: Support `-H:ParseOnce` in Truffle languages

* Main Idea: At analysis time correctly generate and analyze all needed graphs
    - Change analysis to handle multiple graph implementations of a given Java method

* We want to unconditionally use ParseOnce
    - Currently we have many SubstrateOptions.parseOnce() hooks throughout the code
    - ParseOnce needed for Truffle language support in [Spring](https://github.com/oracle/graal/issues/4473)

* I've been adding support throughout the last year
    - New flag option: [`-H:ParseOnceJIT`](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java#L81)

---

## Native Image Runtime Compilation Support Overview

* Main idea: store encoded Graal graphs in the executable
    - At runtime the Partial Evaluator can use them to generate new code

* New runtime-generated code can deoptimize back to code generated at buildtime
    - Destination is called the *Deoptimization Target*

* Generate two versions of methods at buildtime
    - Deoptimization Target
    - Original
    - Do this because Deoptimization Target methods are slower
    - Not discussed: how to optimize Deopt Methods 
        + [Paper](https://dl.acm.org/doi/10.1145/3033019.3033025) written about how it works

---

## Terminology

* Runtime Method: graphs used for runtime compilation
* Deopt(imization) Method: method compiled at buildtime that serves as the deoptimization target for Runtime Methods
* Original/Normal/Regular Method: method compiled at buildtime which is not a deoptimization target
* AOT Method: either Original or Deopt Method
* Deopt Entry: place within the Deopt Method where execution is transferred to when a Runtime Method deoptimizes

---

## Runtime Compilation Example

![Runtime Compilation](https://github.com/oracle/graal/assets/4650103/70383a17-5dd4-4bcd-9f16-36011c16dbfa)

---

## Runtime Deoptimization Example (1)

![Deopt to Method B](https://github.com/oracle/graal/assets/4650103/189e5821-0f5c-485a-8913-e70b01022fcc)

---

## Runtime Deoptimization Example (2)

![Deopt to Method C](https://github.com/oracle/graal/assets/4650103/90f48757-1628-45c6-9df2-7d3b11e59fb3)

---

## Deeper Look: Runtime Compiled Methods

* Compilation at runtime requires storing encoded versions of graphs at buildtime

* During native-image analysis we must determine which methods are candidates for runtime compilation
    - Find all methods reachable from "compilation roots" (methods registered by [prepareMethodForRuntimeCompilation](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/RuntimeCompilationFeature.java#L532))
    - Disregard reachable methods which are not eligible for runtime compilation

* Truffle provides guidelines for methods ineligible for runtime compilation
    - [@TruffleBoundary](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/CompilerDirectives.java#L282)
    - built-in Blocklist
    - More critera in [allowRuntimeCompilation](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleFeature.java#L551)

* Truffle implementations also can provide guidelines on the [Max # of Runtime Compiled Methods](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/RuntimeCompilationFeature.java#L142)
    - Helps to catch Truffle code regressions

* Takeaway: Runtime Compilation is very dependent on analysis to ensure runtime compilation space does not explode

---

## Deeper Look: Deoptimization Targets & Deoptimization Entrypoints

* Deopt methods must serve as the entrypoint for all runtime deopts

* Constraint: Deoptimization can only be from a FrameState with the Runtime method
 
* A [FrameState](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/nodes/FrameState.java#L74) describes a specific bytecode index & state
    - Created during initial graph creation (in [BytecodeParser](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/wasm/src/org.graalvm.wasm/src/org/graalvm/wasm/parser/bytecode/BytecodeParser.java#L78))
    - Used in Graal-JIT mode to deoptimize back to the interpreter
    - Also used in native-image mode to track debug information (via [FullInfoPointNodes](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/nodes/FullInfopointNode.java#L46))

* During runtime graph encoding we record all FrameStates present within the graph 
    - Build Deopt Target graphs to allow Deopt Entries at all these points
    - Validate at image creation with [VerifyDeoptimizationEntryPoints](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageCodeCache.java#L124)

---

## (Quick) Analysis TypeFlow Overview

* During Analysis important nodes in Graal Graphs are represented by [TypeFlows](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/TypeFlow.java#L49) (within [FlowGraphs](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/MethodFlowsGraph.java#L50))

* TypeFlows are linked to their **uses** (link from def &rarr; use)
* A TypeFlow can also **observe** a TypeFlow to receive updates (link from observer &larr; observee)

* Uses and observers form a directed graph of the reachable methods

* Each TypeFlow tracks its [TypeState](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/typestate/TypeState.java#L40) (i.e., types present at that TypeFlow node)

* When a TypeFlow node's TypeState changes, it notifies all of its listeners (uses and observers) of its new TypeState

* When a listener TypeFlow node is notified of a change, it can react to this change
    - Main callbacks: [addState](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/TypeFlow.java#L290), [onObservedUpdate](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/TypeFlow.java#L595)
    - May update its own TypeState
      + When this happens it must also update its listeners

---

## InvokeTypeFlow Linking Overview

* InvokeTypeFlows serve as the connection between different AnalysisMethod FlowGraphs

* On a TypeState update it performs the following tasks;
    1. Determine call targets for new types
    2. Links caller's invoke argument TypeFlows to callee's parameter TypeFlows
    3. Links callee's return TypeFlow to the invoke's TypeFlow uses

*  **Important**: For each new receiver type, linking is performed once
    - Updates are only posted for new receiver types

* More info: see [DefaultStaticInvokeTypeflow](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/typestate/DefaultStaticInvokeTypeFlow.java#L59), [DefaultSpecialInvokeTypeflow](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/typestate/DefaultSpecialInvokeTypeFlow.java#L59), [DefaultVirtualInvokeTypeflow](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/typestate/DefaultVirtualInvokeTypeFlow.java#L57)

---

## ParseOnceJIT Main Tasks

* Introduce Multi-Method Concept
* Selective linking between method type flows
* Dynamically link new methods reachable from runtime compilation (and deoptimization)
* Support inline before analysis

---

## Multi-Method Concept

* One ResolvedJavaMethod can correspond to multiple underlying implementations
    - Original, Deopt, and Runtime methods
    - All have different graph implementations

* [Now](https://github.com/oracle/graal/pull/5212) AnalysisMethod and HostedMethod implement [MultiMethod](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.common/src/com/oracle/svm/common/meta/MultiMethod.java#L35)
    - Each MultiMethod has completely separate analysis and hosted information
    - A given MultiMethod variant is identified by its [MultiMethodKey](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.common/src/com/oracle/svm/common/meta/MultiMethod.java#L54)
        - Also allows for easy navigation between different variants

* Runtime Compilation uses the following keys:
    - [ORIGINAL_METHOD](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.common/src/com/oracle/svm/common/meta/MultiMethod.java#L57)
    - [DEOPT_TARGET_METHOD](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.common/src/com/oracle/svm/common/meta/MultiMethod.java#L64)
    - [RUNTIME_COMPILED_METHOD](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/SubstrateCompilationDirectives.java#L57)

* ORIGINAL_METHOD is "more equal" than other MultiMethodKeys
    - Used more in universe building, image layout

---

## Runtime Compilation Invoke Patterns

* A given Method can "call" (transfer to) a variety of MultiMethods
* &rarr; denotes can call MultiMethod type
* &larr; denotes can be returned to from MultiMethod type

* Runtime
  - &harr; Runtime
  - &harr; Original  
  - &rarr; Deopt  

* Original
  - &harr; Original
  - &harr; Runtime
  - &larr; Deopt

* Deopt 
  - &harr; Original
  - &larr; Deopt
  - &larr; Runtime

---

## MultiMethod Invoke Patterns: Runtime
* &harr; Runtime
* &harr; Original
* &rarr; Deopt

![Deopt to Method C](https://github.com/oracle/graal/assets/4650103/90f48757-1628-45c6-9df2-7d3b11e59fb3)

---

## MultiMethod Invoke Patterns: Original
* &harr; Original
* &harr; Runtime
* &larr; Deopt

![Deopt to Method C](https://github.com/oracle/graal/assets/4650103/90f48757-1628-45c6-9df2-7d3b11e59fb3)

---

## MultiMethod Invoke Patterns: Deopt
* &harr; Original
* &larr; Deopt
* &larr; Runtime

![Deopt to Method C](https://github.com/oracle/graal/assets/4650103/90f48757-1628-45c6-9df2-7d3b11e59fb3)

---

## InvokeTypeFlow Linking

* Added logic to InvokeTypeFlows to allow linking to multiple methods
* Allow parameter and return TypeFlows to be selective linked between caller &harr; callee
* hooks in code: [determineCallees](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L1059), [performReturnLinking](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L1160), [performParameterLinking](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L1146)

---

## Stub flows

* Since linking is performed once, we must eagerly link to runtime and deopt target methods
* However, want to lazily create flows
* Solution: create [STUB](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/MethodFlowsGraph.java#L60) flows
  - Stub flows link to incoming parameters / going uses, but do not have internal flows
  - Created in [MethodTypeFlowBuilder](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/MethodTypeFlowBuilder.java#L598)

* We generate full flow lazily when needed
    - Runtime Method: Once it is reachable from another runtime compiled method (or is a compilation root)
    - Deopt Method: Once it is seen within a Runtime method's FrameState

---

## Updating Deopt Target TypeFlow

* We create deopt entries for every Runtime method's framestate

* Due to inlining, a given method many be present in many different runtime methods' framestates

* Requires us to recreate Deopt graphs and redo internal flow
    - Why recreate? Deopt Entrypoints are inserted during the initial Graal graph generation
        - Nutshell reason: Deopt Entrypoint exception handling logic needs to created at this point

* Added new logic: [updateFlowsGraph](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/MethodTypeFlow.java#L284)

---

## Runtime Method Inlining

* It is important to allow inlining to prevent the number of reachable runtime method from exploding
    - When an invoke saturates it can cause many methods to be reachable for runtime compilation
    - Especially problematic for generic container classes

* Legacy implementation uses [TruffleInlineDuringParsing](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleFeature.java#L501) to guide inlining during parsing
    - Without this most truffle languages can't be built

* We created a new InlineBeforeAnalysis policy for Runtime methods

---

## New Runtime Inlining Policy

* Classify methods as being [trivial to inline](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/InliningUtilities.java#L41)

1. Inline methods up to [InlineDuringParsingMaxDepth](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/java/BytecodeParserOptions.java#L62) (default 10) via this approach
2. If step 1 fails, inline methods according to "regular" policy
    - Inline methods as long as the number of new nodes introduced into the graph is less than [InlineBeforeAnalysisAllowedInvokes](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/InlineBeforeAnalysisPolicyUtils.java#L72) (default 1)

* Code is implemented in [RuntimeCompilationAnalysisPolicy](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L964) and [InlineBeforeAnalysisPolicyUtils](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/InlineBeforeAnalysisPolicyUtils.java#L66)

* No inlining is performed within Deopt methods
    - Calls can serve as implicit Deopt entries
    - Want to exit Deopt methods as quickly as possible

---

## Linking Deopt Methods During Inlining

* Inlining methods prevents typeflows from entering Deopt Targets
    - No longer method boundary where we can hook multiple targets

* Added hooks to insert [InlinedInvokeArgumentsNode](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/nodes/InlinedInvokeArgumentsNode.java#L45) when inlining before analysis
    - Passes parameter values to the Deopt Target, but not other variants

* Does not link the return value
    - Unneeded due to Deopt method's flow being more conservative than Runtime method.

* We create "DeoptInvokeTypeFlows" which link only to the Deopt Method. 

---

## Generating Runtime Graphs

* Runtime methods perform different optimizations than AOT methods
    - No explicit exceptions
    - Allow assumptions
    - Never fold [@CompilationFinal](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/CompilerDirectives.java#L255) (at build-time)
    - additional [GraphBuilderPlugins](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/truffle/compiler/substitutions/TruffleGraphBuilderPlugins.java#L142)

* PE will run during runtime compilation
    - Too much build-time optimization can prevent PE optimizations
        + for example, @CompilationFinal array objects
    - Still beneficial for code size and validation

* Added hooks in analysis create Runtime method graphs correctly: [ParseGraph](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L763)

* Compile & encode runtime graphs in [onCompileQueueCreation](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L650) hook

* Much of the logic for creating the Runtime providers is in TruffleFeature
    - Very unintuitive
    - I will (eventually) clean this up

---

## ParseOnceJIT Current Status

* Most functionality is in `master`
* Enable via adding `-H:+ParseOnceJIT` to build process
    - off by default
* Goal: have `-H:+ParseOnceJIT` as default for GraalVM 23.1
    - will enable soon (upcoming weeks) on `master`
* Improvements will continue for foreseeable future (beyond 23.1)

---

## Code References

* All of these files will undergo major revisions soon

* [ParseOnceRuntimeCompilationFeature](https://github.com/oracle/graal/blob/cd03bb635dbdaefd30b21d58ad291a703db8f341/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/ParseOnceRuntimeCompilationFeature.java#L136)
    - Logic for all ParseOnceJIT-specific functionality
* [RuntimeCompilationFeature](https://github.com/oracle/graal/blob/52d791068b5c34f37b48174500f437aca642f027/substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/hosted/RuntimeCompilationFeature.java#L132)
    - Logic for shared functionality between old (Legacy) & new (ParseOnceJIT) implementations
* [TruffleFeature](https://github.com/oracle/graal/blob/52d791068b5c34f37b48174500f437aca642f027/substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleFeature.java#L188)
    - Creates runtime graph creation environment
    - Provides criteria for whether a method can be runtime-compiled
